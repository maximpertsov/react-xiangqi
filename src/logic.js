import update from 'immutability-helper';

const RANKS = 10;
const FILES = 9;
const BLACK_PIECES = 'rheakcp';
const RED_PIECES = 'RHEAKCP';
const BLACK_RIVER_BANK = 4;
const RED_RIVER_BANK = 5;
const ORTHOGONAL_MOVES = [[1, 0], [-1, 0], [0, 1], [0, -1]];
const DIAGONAL_MOVES = [[1, 1], [-1, 1], [1, -1], [-1, -1]];
// TODO store this in a board FEN-style string?
const BLACK_PALACE = [
  [0, 3], [0, 4], [0, 5],
  [1, 3], [1, 4], [1, 5],
  [2, 3], [2, 4], [2, 5],
];
const RED_PALACE = [
  [9, 3], [9, 4], [9, 5],
  [8, 3], [8, 4], [8, 5],
  [7, 3], [7, 4], [7, 5],
];
const EMPTY_BOARD_FEN = '9/9/9/9/9/9/9/9/9/9';

class XiangqiBoard {
  // TODO can remove most of this information and parse it from the FEN string
  constructor({
    ranks = RANKS,
    files = FILES,
    redPieces = RED_PIECES,
    blackPieces = BLACK_PIECES,
    fen = EMPTY_BOARD_FEN,
    redRiverBank = RED_RIVER_BANK,
    blackRiverBank = BLACK_RIVER_BANK,
    redPalace = RED_PALACE,
    blackPalace = BLACK_PALACE,
  } = {}) {
    this.ranks = ranks;
    this.files = files;
    this.redPieces = redPieces;
    this.blackPieces = blackPieces;
    this.redRiverBank = redRiverBank;
    this.blackRiverBank = blackRiverBank;
    this.redPalace = redPalace.map((pos) => this.getSlot(...pos));
    this.blackPalace = blackPalace.map((pos) => this.getSlot(...pos));
    this.board = this.fromFen(fen);
  }

  move(fromSlot, toSlot) {
    const board = update(update(this.board, {
      [toSlot]: { $set: this.board[fromSlot] },
    }), {
      [fromSlot]: { $set: null },
    });
    return this.new(board, this);
  }

  move(fromSlot, toSlot) {
    const board = update(update(this.board, {
      [toSlot]: { $set: this.board[fromSlot] },
    }), {
      [fromSlot]: { $set: null },
    });
    return this.new(board);
  }

  drop(piece, toSlot) {
    const board = update(this.board, {
      [toSlot]: { $set: piece },
    });
    return this.new(board);
  }

  new(board) {
    const options = { ...this };
    delete options.board;
    options.fen = this.toFen(board);
    options.redPalace = this.redPalace.map((slot) => this.getRankFile(slot));
    options.blackPalace = this.blackPalace.map((slot) => this.getRankFile(slot));
    return new this.constructor(options);
  }

  getPiece(rank, file) { return this.board[this.getSlot(rank, file)]; }

  getSlot(rank, file) { return file + rank * this.files; }

  getRank(slot) { return Math.floor(slot / this.files); }

  getFile(slot) { return slot % this.files; }

  getRankFile(slot) { return [this.getRank(slot), this.getFile(slot)]; }

  isRed(code) { return this.redPieces.includes(code); }

  isBlack(code) { return this.blackPieces.includes(code); }

  sameColor(code1, code2) {
    return (this.isRed(code1) && this.isRed(code2))
      || (this.isBlack(code1) && this.isBlack(code2));
  }

  // TODO consolidate with sameColor method
  sameColorSlot(slot1, slot2) {
    const code1 = this.board[slot1];
    const code2 = this.board[slot2];
    return this.sameColor(code1, code2);
  }

  static fromFenRow(row) {
    return row.split('').reduce((acc, ch) => {
      const val = +ch;
      const newItems = Number.isNaN(val) ? [ch] : Array(val).fill(null);
      return acc.concat(newItems);
    }, []);
  }

  fromFen(fen) {
    return fen.split('/').reduce(
      (acc, row) => acc.concat(this.constructor.fromFenRow(row)),
      [],
    );
  }

  getNextRankSlot(slot) {
    const code = this.board[slot];
    const rank = this.getRank(slot);
    const file = this.getFile(slot);
    let nextRank = rank;
    if (this.isBlack(code)) nextRank = Math.min(rank + 1, this.ranks - 1);
    if (this.isRed(code)) nextRank = Math.max(rank - 1, 0);
    return this.getSlot(nextRank, file);
  }

  crossingRiver(fromSlot, toSlot) {
    const code = this.board[fromSlot];
    const rank = this.getRank(toSlot);
    if (this.isBlack(code)) return rank >= this.redRiverBank;
    if (this.isRed(code)) return rank <= this.blackRiverBank;
    return false;
  }

  isUniverallyLegal(fromSlot, toSlot) {
    if (toSlot === null) return false;
    if (fromSlot === toSlot) return false;
    if (this.sameColor(this.board[fromSlot], this.board[toSlot])) return false;
    return true;
  }

  addIfUniversallyLegal(moves, fromSlot, toSlot) {
    if (this.isUniverallyLegal(fromSlot, toSlot)) {
      moves.push(toSlot);
    }
  }

  tryMove(slot, rankMove, fileMove) {
    const [rank, file] = this.getRankFile(slot);
    const newRank = rankMove + rank;
    const newFile = fileMove + file;
    if (newRank < 0 || newRank >= this.ranks) return null;
    if (newFile < 0 || newFile >= this.files) return null;
    return this.getSlot(newRank, newFile);
  }

  tryMoves(slot, moves) {
    return moves.map((m) => this.tryMove(slot, ...m));
  }

  // TODO make this non-recursive?
  tryMarch(slot, rankMove, fileMove, steps = Math.max(this.ranks, this.files)) {
    if (steps < 1) return [];
    const nextSlot = this.tryMove(slot, rankMove, fileMove);
    if (nextSlot === null) return [];
    const restSlots = this.tryMarch(nextSlot, rankMove, fileMove, steps - 1);
    restSlots.push(nextSlot);
    return restSlots;
  }

  tryMarchMoves(slot, moves, steps) {
    return moves.reduce(
      (acc, move) => acc.concat(this.tryMarch(slot, move[0], move[1], steps)),
      [],
    );
  }

  legalPawnMoves(slot) {
    const result = [];
    const forwardSlot = this.getNextRankSlot(slot);
    this.addIfUniversallyLegal(result, slot, forwardSlot);
    if (this.crossingRiver(slot, slot)) {
      this.addIfUniversallyLegal(result, slot, this.tryMove(slot, 0, -1));
      this.addIfUniversallyLegal(result, slot, this.tryMove(slot, 0, 1));
    }
    return result;
  }

  orthogonalSlots(slot, radius) {
    const steps = radius === undefined ? 1 : radius;
    return this.tryMarchMoves(slot, ORTHOGONAL_MOVES, steps);
  }

  diagonalSlots(slot, radius) {
    const steps = radius === undefined ? 1 : radius;
    return this.tryMarchMoves(slot, DIAGONAL_MOVES, steps);
  }

  isOccupied(slot) {
    return this.board[slot] !== null;
  }

  legalHorseMoves(slot) {
    const result = [];

    this.orthogonalSlots(slot).forEach((firstHop, _, firstHops) => {
      if (this.isOccupied(firstHop)) return;

      this.diagonalSlots(firstHop).forEach((secondHop) => {
        if (firstHops.includes(secondHop) || result.includes(secondHop)) return;
        this.addIfUniversallyLegal(result, slot, secondHop);
      });
    });

    return result;
  }

  legalRookMoves(slot) {
    const result = [];
    ORTHOGONAL_MOVES.forEach((move) => {
      let toSlot = slot;
      while (true) {
        toSlot = this.tryMove(toSlot, ...move);
        if (toSlot === null) break;
        this.addIfUniversallyLegal(result, slot, toSlot);
        if (this.isOccupied(toSlot)) break;
      }
    });
    return result;
  }

  legalCannonMoves(slot) {
    const result = [];
    ORTHOGONAL_MOVES.forEach((move) => {
      let toSlot = slot;
      let vaulted = false;
      while (true) {
        toSlot = this.tryMove(toSlot, ...move);
        if (toSlot === null) break;
        if (vaulted && this.isOccupied(toSlot)) {
          this.addIfUniversallyLegal(result, slot, toSlot);
          break;
        } else if (this.isOccupied(toSlot)) {
          vaulted = true;
        } else if (!vaulted) {
          this.addIfUniversallyLegal(result, slot, toSlot);
        }
      }
    });
    return result;
  }

  legalElephantMoves(slot) {
    const result = [];
    DIAGONAL_MOVES.forEach((move) => {
      const firstHop = this.tryMove(slot, ...move);
      if (this.isOccupied(firstHop) || this.crossingRiver(slot, firstHop)) {
        return;
      }

      const secondHop = this.tryMove(firstHop, ...move);
      this.addIfUniversallyLegal(result, slot, secondHop);
    });
    return result;
  }

  inPalace(fromSlot, toSlot) {
    const code = this.board[fromSlot];
    if (this.isBlack(code)) return this.blackPalace.includes(toSlot);
    if (this.isRed(code)) return this.redPalace.includes(toSlot);
    return false;
  }

  legalAdvisorMoves(slot) {
    const result = [];
    this.diagonalSlots(slot, 1).filter((s) => this.inPalace(slot, s)).forEach((s) => {
      this.addIfUniversallyLegal(result, slot, s);
    });
    return result;
  }

  legalKingMoves(slot) {
    const result = [];
    this.orthogonalSlots(slot, 1).filter((s) => this.inPalace(slot, s)).forEach((s) => {
      this.addIfUniversallyLegal(result, slot, s);
    });
    return result;
  }

  // TODO should not allow self-checks by default
  legalMoves(allowSelfCheck) {
    const result = this.board.map((code, slot) => {
      if (code === 'p' || code === 'P') return this.legalPawnMoves(slot);
      if (code === 'h' || code === 'H') return this.legalHorseMoves(slot);
      if (code === 'r' || code === 'R') return this.legalRookMoves(slot);
      if (code === 'c' || code === 'C') return this.legalCannonMoves(slot);
      if (code === 'e' || code === 'E') return this.legalElephantMoves(slot);
      if (code === 'a' || code === 'A') return this.legalAdvisorMoves(slot);
      if (code === 'k' || code === 'K') return this.legalKingMoves(slot);
      return [];
    });

    if (allowSelfCheck) return result;

    return result.map((toSlots, fromSlot) => (
      toSlots.filter((toSlot) => (!this.checksOwnKing(fromSlot, toSlot)))
    ));
  }

  captures() {
    const result = new Set();
    for (const [_fromSlot, toSlots] of this.legalMoves(true).entries()) {
      toSlots.forEach((slot) => {
        if (this.isOccupied(slot)) result.add(this.board[slot]);
      });
    }
    return result;
  }

  // TODO add check for two kings facing each other
  //      hack this by replacing the opposing king with a rook?
  checksOwnKing(fromSlot, toSlot) {
    const code = this.board[fromSlot];
    let ownKing;
    let otherKing;
    let otherRook;
    if (this.isBlack(code)) [ownKing, otherKing, otherRook] = ['k', 'K', 'R'];
    if (this.isRed(code)) [ownKing, otherKing, otherRook] = ['K', 'k', 'r'];

    const nextBoard = this.move(fromSlot, toSlot);
    return nextBoard.drop(
      otherRook,
      nextBoard.board.indexOf(otherKing)
    ).captures().has(ownKing);
  }

  toFen(board = this.board) {
    const rows = [];
    board.forEach((piece, idx) => {
      if (idx % this.files === 0) rows.push([]);
      const lastRow = rows[rows.length - 1];
      const lastRowSize = lastRow.length;
      if (piece !== null) {
        lastRow.push(piece);
      } else if (lastRowSize === 0 || Number.isNaN(+lastRow[lastRowSize - 1])) {
        lastRow.push(1);
      } else {
        lastRow[lastRowSize - 1]++;
      }
    });
    return rows.map((row) => row.join('')).join('/');
  }
}

export default XiangqiBoard;
