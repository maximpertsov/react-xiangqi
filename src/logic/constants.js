export const RefType = Object.freeze({
  SLOT: 0,
  RANK_FILE: 1,
  RANK_FILE_STRING: 2,
});

export const Color = Object.freeze({
  RED: 'red',
  BLACK: 'black',
});

export const Piece = Object.freeze({
  Black: Object.freeze({
    CHARIOT: 'r',
    HORSE: 'h',
    ELEPHANT: 'e',
    ADVISOR: 'a',
    GENERAL: 'k',
    CANNON: 'c',
    PAWN: 'p',
  }),
  Red: Object.freeze({
    CHARIOT: 'R',
    HORSE: 'H',
    ELEPHANT: 'E',
    ADVISOR: 'A',
    GENERAL: 'K',
    CANNON: 'C',
    PAWN: 'P',
  }),
});

export const BLACK_PIECES = Object.values(Piece.Black);
export const RED_PIECES = Object.values(Piece.Red);
export const RANK_COUNT = 10;
export const FILE_COUNT = 9;
export const BLACK_RIVER_BANK = 4;
export const RED_RIVER_BANK = 5;
export const ORTHOGONAL_MOVES = [[1, 0], [-1, 0], [0, 1], [0, -1]];
export const DIAGONAL_MOVES = [[1, 1], [-1, 1], [1, -1], [-1, -1]];
