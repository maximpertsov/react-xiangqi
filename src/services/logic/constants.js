export const Color = Object.freeze({
  RED: 'red',
  BLACK: 'black',
});

export const Piece = Object.freeze({
  Black: Object.freeze({
    CHARIOT: 'r',
    HORSE: 'n',
    ELEPHANT: 'b',
    ADVISOR: 'a',
    GENERAL: 'k',
    CANNON: 'c',
    PAWN: 'p',
  }),
  Red: Object.freeze({
    CHARIOT: 'R',
    HORSE: 'N',
    ELEPHANT: 'B',
    ADVISOR: 'A',
    GENERAL: 'K',
    CANNON: 'C',
    PAWN: 'P',
  }),
});

export const BLACK_PIECES = Object.values(Piece.Black);
export const RED_PIECES = Object.values(Piece.Red);
export const RANK_COUNT = 10;
export const FILE_COUNT = 9;
export const BLACK_RIVER_BANK = 4;
export const RED_RIVER_BANK = 5;
export const ORTHOGONAL_MOVES = [[1, 0], [-1, 0], [0, 1], [0, -1]];
export const DIAGONAL_MOVES = [[1, 1], [-1, 1], [1, -1], [-1, -1]];
export const BLACK_PALACE = [
  [0, 3], [0, 4], [0, 5],
  [1, 3], [1, 4], [1, 5],
  [2, 3], [2, 4], [2, 5],
];
export const RED_PALACE = [
  [9, 3], [9, 4], [9, 5],
  [8, 3], [8, 4], [8, 5],
  [7, 3], [7, 4], [7, 5],
];
