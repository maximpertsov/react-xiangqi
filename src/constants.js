import { Color } from 'services/logic/constants';

export const AutoMove = Object.freeze({
  BLACK: Color.BLACK,
  BOTH: 'both',
  OFF: 'off',
  RED: Color.RED,
});

export default {};
