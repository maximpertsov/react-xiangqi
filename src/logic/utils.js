import { Piece, FILE_COUNT } from './constants';

export const isPawn = (piece) => (
  piece === Piece.Black.PAWN || piece === Piece.Red.PAWN
);
export const isChariot = (piece) => (
  piece === Piece.Black.CHARIOT || piece === Piece.Red.CHARIOT
);
export const isHorse = (piece) => (
  piece === Piece.Black.HORSE || piece === Piece.Red.HORSE
);
export const isElephant = (piece) => (
  piece === Piece.Black.ELEPHANT || piece === Piece.Red.ELEPHANT
);
export const isGeneral = (piece) => (
  piece === Piece.Black.GENERAL || piece === Piece.Red.GENERAL
);
export const isCannon = (piece) => (
  piece === Piece.Black.CANNON || piece === Piece.Red.CANNON
);
export const isAdvisor = (piece) => (
  piece === Piece.Black.ADVISOR || piece === Piece.Red.ADVISOR
);

export const getSlot = (rank, file) => file + (rank * FILE_COUNT);
export const getRank = (slot) => Math.floor(slot / FILE_COUNT);
export const getFile = (slot) => slot % FILE_COUNT;
export const getRankFile = (slot) => [getRank(slot), getFile(slot)];
