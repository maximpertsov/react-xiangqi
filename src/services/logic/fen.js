import zipObject from 'lodash/zipObject';
import update from 'lodash/update';
import { Color } from 'service/logic/constants'

const FEN_FIELDS = [
  'placement',
  'activeColor',
  'castling',
  'enPassant',
  'halfmoves',
  'fullmoves',
];

const activeColor = (symbol) => {
  switch (symbol) {
    case 'w':
      return Color.RED
    case 'b':
      return Color.BLACK
    default:
      throw new Error(`Invalid piece color ${symbol}`);
  }

const decodeFields = fen => {
  const result = zipObject(FEN_FIELDS, fen.split(' '));
  update(result, 'activeColor', activeColor);
  update(result, 'halfmoves', parseInt);
  update(result, 'fullmoves', parseInt);
  return result;
};

const decodeRow = row =>
  row.split('').reduce((acc, ch) => {
    const val = +ch;
    const newItems = Number.isNaN(val) ? [ch] : Array(val).fill(null);
    return acc.concat(newItems);
  }, []);

const decodePlacement = placement =>
  placement.split('/').reduce((acc, row) => acc.concat(decodeRow(row)), []);

export const decode = fen => {
  const result = decodeFields(fen);
  update(result, 'placement', decodePlacement);
  return result;
};

export default {};
