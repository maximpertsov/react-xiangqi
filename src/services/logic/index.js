import { Color } from './constants';
import { sameColor } from './utils';
import { decode as decodeFen } from './fen';
import { makeMove } from './move';
import { decode as decodeSquare, encode as encodeSquare } from './square';

export default class XiangqiBoard {
  constructor({ fen, ...options }) {
    const params = fen === undefined ? options : decodeFen(fen);
    this.placement = params.placement;
    this.activeColor = params.activeColor;
  }

  // TODO: keep and test
  move(move) {
    return this.new({ placement: makeMove(this.placement, move) });
  }

  new(options) {
    return new this.constructor(
      options === undefined ? { ...this } : { ...options },
    );
  }

  getPiece(square) {
    return this.placement[decodeSquare(square)];
  }

  sameColor(square1, square2) {
    const slot1 = decodeSquare(square1);
    const slot2 = decodeSquare(square2);
    const piece1 = this.placement[slot1];
    const piece2 = this.placement[slot2];

    return sameColor(piece1, piece2);
  }

  isOccupied(square) {
    const slot = decodeSquare(square);

    return this.placement[slot] !== null;
  }

  activeKing() {
    let king = undefined;

    switch (this.activeColor) {
      case Color.RED:
        king = 'K';
        break;
      case Color.BLACK:
        king = 'k';
        break;
      default:
        // TODO: throw error
        return;
    }

    const kingSlot = this.placement.indexOf(king);
    return encodeSquare(kingSlot);
  }
}
