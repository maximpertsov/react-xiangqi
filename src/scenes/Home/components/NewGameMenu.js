import React, { useContext } from 'react';
import { Button, Header, Icon, Popup } from 'semantic-ui-react';
import { useSelector } from 'react-redux';
import { createSelector } from 'reselect';
import isEqual from 'lodash/isEqual';

import { Team } from 'services/logic/constants';
import client from 'services/client';
import { WebSocketContext } from 'services/WebSocketProvider';

import find from 'lodash/find';
import styled from '@emotion/styled';

const Wrapper = styled.div`
  padding: 5px;
`;

const mapStateToProps = createSelector(
  [state => state],

  state => ({
    username: state.username,
    ownLobbyRequest: find(state.lobbyGames, { player1: state.username }),
  }),
);

const NewGameMenu = () => {
  const io = useContext(WebSocketContext);

  const { username, ownLobbyRequest } = useSelector(mapStateToProps, isEqual);

  const createGameRequest = team => () => {
    client
      .post('game/request', {
        player1: username,
        parameters: { team },
      })
      .then(() => {
        io.send({ type: 'updated_lobby_games' });
      });
  };

  const cancelGameRequest = id => () => {
    client.delete(`game/request/${id}`).then(() => {
      io.send({ type: 'updated_lobby_games' });
    });
  };

  const renderRequestButtons = () => (
    <Button.Group fluid>
      <Popup
        content="Play with red pieces"
        trigger={
          <Button icon onClick={createGameRequest(Team.RED)}>
            <Icon color="red" fitted name="plus square outline" />
          </Button>
        }
      />
      <Popup
        content="Play with black pieces"
        trigger={
          <Button icon onClick={createGameRequest(Team.BLACK)}>
            <Icon color="black" fitted name="plus square outline" />
          </Button>
        }
      />
      <Popup
        content="Play with random pieces"
        trigger={
          <Button icon onClick={createGameRequest()}>
            <Icon fitted name="question" />
          </Button>
        }
      />
    </Button.Group>
  );

  const renderCancelButton = () => (
    <Button onClick={cancelGameRequest(ownLobbyRequest.id)} fluid loading />
  );

  return (
    <Wrapper className="NewGameMenu">
      <Header size="medium">Create game</Header>
      {ownLobbyRequest ? renderCancelButton() : renderRequestButtons()}
    </Wrapper>
  );
};

export default NewGameMenu;
