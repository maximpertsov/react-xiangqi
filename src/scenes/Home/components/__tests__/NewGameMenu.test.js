import React from 'react';
import { useSelector } from 'react-redux';
import axios from 'axios';
import { Team } from 'services/logic/constants';

import NewGameMenu from '../NewGameMenu';

jest.mock('react-redux');
jest.mock('axios');

describe('NewGameMenu', () => {
  const username = 'alice';
  let store;
  let wrapper;

  beforeEach(() => {
    useSelector.mockImplementation(callback => callback(store.getState()));
    wrapper = shallowWrappedComponent(<NewGameMenu />, store);
  });

  afterEach(() => {
    store.clearActions();
    jest.resetAllMocks();
  });

  describe('did not make game request', () => {
    beforeAll(() => {
      store = mockStore({
        lobbyGames: [{ id: 123, player1: 'bob', parameters: {} }],
        username,
      });
    });

    beforeEach(() => {
      axios.post.mockResolvedValue({});
    });

    test('snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    describe('click', () => {
      const table = [
        ['Play with red pieces', { team: Team.RED }],
        ['Play with black pieces', { team: Team.BLACK }],
        ['Play with random pieces', { team: undefined }],
      ];

      test.each(table)('%i', (content, parameters) => {
        expect.assertions(1);
        const button = wrapper.wrap(
          wrapper
            .find('Popup')
            .findWhere(node => node.prop('content') === content)
            .prop('trigger'),
        );
        button.simulate('click');

        expect(axios.post).toHaveBeenCalledWith('game/request', {
          player1: username,
          parameters,
        });
      });
    });
  });

  describe('already made game request', () => {
    const id = 789;

    beforeAll(() => {
      store = mockStore({
        lobbyGames: [{ id, player1: username, parameters: {} }],
        username,
      });
    });

    beforeEach(() => {
      axios.delete.mockResolvedValue({});
    });

    test('snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('click', () => {
      expect.assertions(1);
      wrapper.find('Button').simulate('click');

      expect(axios.delete).toHaveBeenCalledWith(`game/request/${id}`);
    });
  });
});
