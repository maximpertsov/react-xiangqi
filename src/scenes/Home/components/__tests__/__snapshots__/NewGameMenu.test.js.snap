// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewGameMenu already made game request snapshot 1`] = `
<Styled(div)
  className="NewGameMenu"
>
  <Header
    size="medium"
  >
    Create game
  </Header>
  <Styled(div)
    repeat={1}
  >
    <Button
      as="button"
      icon={true}
      loading={true}
      onClick={[Function]}
    >
      <Icon
        as="i"
      />
    </Button>
  </Styled(div)>
</Styled(div)>
`;

exports[`NewGameMenu did not make game request snapshot 1`] = `
<Styled(div)
  className="NewGameMenu"
>
  <Header
    size="medium"
  >
    Create game
  </Header>
  <Styled(div)
    repeat={3}
  >
    <Popup
      content="Play with red pieces"
      disabled={false}
      eventsEnabled={true}
      offset={0}
      on={
        Array [
          "click",
          "hover",
        ]
      }
      pinned={false}
      position="top left"
      trigger={
        <Button
          as="button"
          icon={true}
          onClick={[Function]}
        >
          <Icon
            as="i"
            color="red"
            fitted={true}
            name="plus square outline"
          />
        </Button>
      }
    />
    <Popup
      content="Play with black pieces"
      disabled={false}
      eventsEnabled={true}
      offset={0}
      on={
        Array [
          "click",
          "hover",
        ]
      }
      pinned={false}
      position="top left"
      trigger={
        <Button
          as="button"
          icon={true}
          onClick={[Function]}
        >
          <Icon
            as="i"
            color="black"
            fitted={true}
            name="plus square outline"
          />
        </Button>
      }
    />
    <Popup
      content="Play with random pieces"
      disabled={false}
      eventsEnabled={true}
      offset={0}
      on={
        Array [
          "click",
          "hover",
        ]
      }
      pinned={false}
      position="top left"
      trigger={
        <Button
          as="button"
          icon={true}
          onClick={[Function]}
        >
          <Icon
            as="i"
            fitted={true}
            name="question"
          />
        </Button>
      }
    />
  </Styled(div)>
</Styled(div)>
`;
